### DISCUSSION TITLE
Source generator - how to reference a project?

### DISCUSSION THREAD

**Discussion by hugo-the-first** [dotnet/roslyn#48746] [Q&A]
I have a source generator project (SourceGeneratorSamples.csproj) from where I want to reference another project (MyLib.csproj) in the same solution. This lib is needed to generate the source text AND is also referenced inside that generated code. 
Here the SourceGeneratorSamples.csproj:
```xml
<Project Sdk="Microsoft.NET.Sdk">
	<PropertyGroup>
		<TargetFramework>net5.0</TargetFramework>
		<LangVersion>preview</LangVersion>
	</PropertyGroup>
	<ItemGroup>
		<PackageReference Include="Microsoft.CodeAnalysis.CSharp" Version="3.8.0-4.final" PrivateAssets="all" />
		<PackageReference Include="Microsoft.CodeAnalysis.Analyzers" Version="3.3.0" PrivateAssets="all" />
	</ItemGroup>
	<ItemGroup>
		<ProjectReference Include="..\MyLib\MyLib.csproj" />
	</ItemGroup>
</Project>
```

When I try to use this generator in a simple console program (GeneratedDemo) I get a warning and an error.

CSC : warning CS8785: Generator 'ModelGenerator' failed to generate source. It will not contribute to the output and compilation errors may occur as a result. Exception was of type 'FileNotFoundException' with message 'Could not load file or assembly 'MyLib, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null'. The system cannot find the file specified.'

C:\Users\xxxx\source\repos\SourceGenerators\GeneratedDemo\Program.cs(5,7): error CS0246: The type or namespace name 'ModelGenerated' could not be found (are you missing a using directive or an assembly reference?) [C:\Users\xxxx\source\repos\SourceGenerators\GeneratedDemo\GeneratedDemo.csproj]

and here the GeneratedDemo.csproj:

```xml
<Project Sdk="Microsoft.NET.Sdk">
	<PropertyGroup>
		<OutputType>Exe</OutputType>
		<TargetFramework>net5.0</TargetFramework>
		<LangVersion>preview</LangVersion>
	</PropertyGroup>
	<ItemGroup>
		<ProjectReference Include="..\SourceGeneratorSamples\SourceGeneratorSamples.csproj" OutputItemType="Analyzer" ReferenceOutputAssembly="false" />
	</ItemGroup>
</Project>
```
I tried everything - PrivateAssets, GeneratePackageOnBuild, IncludeBuildOutput, DevelopmentDependency, GeneratePathProperty, TargetPathWithTargetPlatformMoniker . . . this is so painful

What's wrong?

**Comment by Shayan-To** [comment]
This has been asked [before](https://github.com/dotnet/roslyn/discussions/47517). But the accepted answered by @sharwell is not about project references, it is about package references.

Take a look, maybe you can work something out.

(I am having trouble having a project reference as well.)

**Comment by sven-n** [comment]
I had the same problem but found a workaround.

In SourceGeneratorSamples.csproj, replace
```
	<ItemGroup>
		<ProjectReference Include="..\MyLib\MyLib.csproj" />
	</ItemGroup>
```
with a reference to the built dll (your HintPath might need to be changed):
```
  <ItemGroup>
    <Reference Include="MyLib">
      <HintPath>..\MyLib\bin\$(Configuration)\MyLib.dll</HintPath>
    </Reference>
  </ItemGroup>
```

Next, we need to make sure that the dll is always built before the generator is built:
1. In the solution explorer, make a right click on the solution
2. Select 'Project Dependencies...'
3. At the combobox, select your generator project
4. Check 'MyLib'
5. OK

I didn't get it to work to build under linux / linux docker container, though :-(

**Reply by OronDF343** [reply]
This does not work in VS 2022

**Reply by TheBoxyBear** [reply]
The option was renamed to `Project Build Dependencies`

**Comment by onetsmr** [comment]
I have such generator project file and all is ok. I can use classes from UIEngine.Vue and UIEngine projects.
```
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <TargetFramework>netstandard2.0</TargetFramework>
    <LangVersion>preview</LangVersion>
    <UserSecretsId>8a900553-ee3a-4e81-aee3-f4eda6a14256</UserSecretsId>
    <GeneratePackageOnBuild>true</GeneratePackageOnBuild>
    <IncludeBuildOutput>false</IncludeBuildOutput>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="Microsoft.CodeAnalysis.CSharp" Version="3.8.0" PrivateAssets="all" />
    <PackageReference Include="Microsoft.CodeAnalysis.Analyzers" Version="3.3.2" PrivateAssets="all" />
  </ItemGroup>

  <ItemGroup>
    <ProjectReference Include="..\UIEngine.Vue\UIEngine.Vue.csproj" />
    <ProjectReference Include="..\UIEngine\UIEngine.csproj" />
  </ItemGroup>

</Project>
```

My main project that uses generator

```
<Project Sdk="Microsoft.NET.Sdk.Web">

  <PropertyGroup>
    <TargetFramework>net5.0</TargetFramework>
    <EmitCompilerGeneratedFiles>true</EmitCompilerGeneratedFiles>
    <CompilerGeneratedFilesOutputPath>$(BaseIntermediateOutputPath)\GeneratedFiles</CompilerGeneratedFilesOutputPath>
  </PropertyGroup>

  ... missing ...

  <ItemGroup>
    ... missing ...
    
    <ProjectReference Include="..\UIEngine.Vue.Generator\UIEngine.Vue.Generator.csproj" OutputItemType="Analyzer" ReferenceOutputAssembly="false" />
  </ItemGroup>
  
  ... missing ...

</Project>
```

My IDE
```
Microsoft Visual Studio Community 2019
Version 16.8.4
VisualStudio.16.Release/16.8.4+30907.101
Microsoft .NET Framework
Version 4.8.04084
```

**Reply by user72356** [reply]
Correct, you have to create a project reference of the shared/common project in the main consuming project (the one that triggers the source generator). This will bring the shared project assembly over to the same space where the generator runs. Note that you still have to create a regular project reference to the shared project within the generator itself. It's kinda lame but that's the "official" way of doing it. It has been commented on by an MS employee at https://github.com/dotnet/roslyn/discussions/47517#discussioncomment-63842

**Reply by laicasaane** [reply]
Hi. Thanks for the explanation. However I hope you would be kind enough to make an example and put it on github. Please?

**Comment by onetsmr** [comment]
Now I have such problem in 16.9.0 - 16.9.2 versions of VS2019. But at the same time nuget reference works fine.

I've added one of the assembly as project reference into the generator project. And sometimes I have 
`Exception was of type 'FileNotFoundException' with message 'Could not load file or assembly [that lib].dll'`

But sometimes it works! What I do? Just comment/uncomment that code line in the Initialize method:
`if (System.Diagnostics.Debugger.IsAttached == false) { System.Diagnostics.Debugger.Launch(); }`
Then it works 1-2-3-5 times.

I also noticed that I have exception when current directory is C:/Program Files.
I have not exception when current directory is my solution/root project directory.
But it doesn't depend on me. VS2019 independently chooses in which directory to run the generator.

How ask VS2019 run generator in the project directory always?
Or how tell VS2019 where this project reference dll?

There was not in the 16.8.xx versions of VS2019.

**Reply by onetsmr** [reply]
One more observation. I have that error when I build in Debug. If I switch to Release and build then error goes away. If I switch back to Debug I have not error. VS2019 16.9.4. This is repo with my experimental code https://github.com/onetsmr/NetGenerator5 which reproduces the error.

**Reply by Sohra** [reply]
Thank you, this tip switching to a Release build helped me with the same issue on VS 16.10.4.

**Comment by laicasaane** [comment]
I met the same problem and tried anything I could. Then my choice was just giving it up atm. This just need a proper fix. I hope the issue can be resolved at the time .NET 6 is released.

**Reply by andres-m-rodriguez** [reply]
I am having the same issue in 2025.... still no fix
